GitUI/UserControls/RevisionGrid/RevisionGridControl.cs

```diff
@@ -914,40 +916,61 @@ public void PerformRefreshRevisions(Func<RefsFilter, IReadOnlyList<IGitRef>> get
                 _gridView.SelectionChanged += OnGridViewSelectionChanged;
                 _gridView.ResumeLayout();
 
-                string pathFilter = BuildPathFilter(_filterInfo.PathFilter);
-                _revisionReader.Execute(
-                    Module,
-                    refs,
-                    revisions,
-                    _filterInfo.CommitsLimit,
-                    _filterInfo.RefFilterOptions,
-                    _filterInfo.IsShowFilteredBranchesChecked ? _filterInfo.BranchFilter : string.Empty,
-                    _filterInfo.GetRevisionFilter(),
-                    pathFilter);
-
-                if (_initialLoad)
-                {
-                    _selectionTimer.Enabled = false;
-                    _selectionTimer.Stop();
-                    _selectionTimer.Enabled = true;
-                    _selectionTimer.Start();
-
-                    _initialLoad = false;
-                }
+                // Add back and show the spinner control, which was removed by SetPage call
+                Controls.Add(_loadingControlSync);
+                ShowLoading();
 
-                _superprojectCurrentCheckout = null;
                 ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
                 {
-                    var scc = await GetSuperprojectCheckoutAsync(capturedModule, noLocks: true);
-                    await this.SwitchToMainThreadAsync();
-
-                    if (_superprojectCurrentCheckout != scc)
+                    cancellationToken.ThrowIfCancellationRequested();
+                    await TaskScheduler.Default;
+
+                    // Find all ambiguous refs (including stash, notes etc)
+                    // NB: This can be EXTREMELY slow operation for large repos.
+                    IReadOnlyList<IGitRef> refs = (getRefs ?? Module.GetRefs)(RefsFilter.NoFilter);
+                    _ambiguousRefs = new(() => GitRef.GetAmbiguousRefNames(refs));
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    string pathFilter = BuildPathFilter(_filterInfo.PathFilter);
+                    _revisionReader.Execute(
```

Comment:
cancellationtoken shold be propagated to RevisionReader (an internal token is used now)
I tried this without seeing any difference though

Reply:
I had that thought too. The CT would need to be [linked](https://docs.microsoft.com/dotnet/standard/threading/how-to-listen-for-multiple-cancellation-requests) because we could be different source of cancellation - both in the revision grid and the revision reader.
Right now we have several layers of CTs, that can be considered somewhat excessive but work I left them as-is, as this change will require significantly more testing than I can afford atm.

Reply:
Revision Reader has RunAsync, this should be changed if revgrid is running in a background task.
(That should be simple, I can push an update.)

Reply:
I'd appreciate it, thanks